agent/src/beerocks/slave/son_slave_thread.cpp

```diff
@@ -4323,21 +4348,31 @@ bool slave_thread::handle_autoconfiguration_wsc(Socket *sd, ieee1905_1::CmduMess
     }
 
     for (auto tlvWscM2 : m2_list) {
+        uint8_t authkey[32];
+        uint8_t keywrapkey[16];
+        if (!autoconfig_wsc_calculate_keys(tlvWscM2, authkey, keywrapkey))
+            return false;
+        LOG(DEBUG) << "M2 Parse: calculate keys";
         auto encrypted_settings = tlvWscM2->encrypted_settings();
-        auto ssid  = std::string(encrypted_settings->ssid(), encrypted_settings->ssid_length());
-        auto bssid = network_utils::mac_to_string(encrypted_settings->bssid_attr().data);
-        auto authentication_type = encrypted_settings->authentication_type_attr().data;
-        auto encryption_type     = encrypted_settings->encryption_type_attr().data;
-        // TODO - finish with decryption and authentication
-        // https://github.com/prplfoundation/prplMesh/pull/91
-        // char network_key[WSC::WSC_MAX_NETWORK_KEY_LENGTH];
-        // std::copy_n(encrypted_settings.network_key_attr().data,
-        //             encrypted_settings.network_key_attr().data_length,
-        //             network_key);
-        // char key_wrap_auth[WSC::WSC_KEY_WRAP_AUTH_LENGTH];
-        // std::copy_n(encrypted_settings.key_wrap_auth_attr().data,
-        //             encrypted_settings.key_wrap_auth_attr().data_length,
-        //             network_key);
+        uint8_t *iv             = reinterpret_cast<uint8_t *>(encrypted_settings->iv());
+        LOG(DEBUG) << "M2 Parse: aes decrypt";
+        mapf::encryption::aes_decrypt(keywrapkey, iv,
+                                      (uint8_t *)encrypted_settings->encrypted_settings(),
+                                      encrypted_settings->encrypted_settings_length());
+
+        // TODO authenticate encrypted settings
+        // before passing to the config_data constructor
+        uint8_t buf[encrypted_settings->encrypted_settings_length()];
+        std::copy_n(encrypted_settings->encrypted_settings(),
+                    encrypted_settings->encrypted_settings_length(), buf);
+        LOG(DEBUG) << "M2 Parse: parse config_data, len = " << sizeof(buf);
```

Comment:
Why do we need to make a copy here? We could just construct config_data with encrypted_settings() as buffer, no?

If we do need to make a copy, then actually it would be better to change aes_decrypt so it *doesn't* work in-place.

Reply:
Will change it, you're right.